llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -19945,6 +19945,41 @@ static bool checkTreeSizes(ArrayRef<std::pair<unsigned, unsigned>> Sizes,
   return Dev * 96 / (Mean * Mean) == 0;
 }
 
+namespace {
+
+/// A group of instructions that we'll try to bundle together using vector ops.
+/// They are ordered using the signed distance of their address operand to the
+/// address of this group's BaseInstr.
+struct RelatedStoreInsts {
+  RelatedStoreInsts(unsigned BaseInstrIdx) { reset(BaseInstrIdx); }
+  void reset(unsigned NewBaseInstr) {
+    BaseInstrIdx = NewBaseInstr;
+    Instrs.clear();
+    insert(NewBaseInstr, 0);
+  }
+
+  void insert(unsigned InstrIdx, int PtrDist) {
+    Instrs.emplace(PtrDist, InstrIdx);
+  }
+
+  /// Return the instruction with a pointer distance of \p PtrDist, or nullopt.
+  std::optional<unsigned> getInstIdx(int PtrDist) const {
+    auto It = Instrs.find(PtrDist);
+    if (It != Instrs.end())
+      return It->second;
+    return std::nullopt;
+  }
```

Comment:
Given the only use of this is where alexey would like to avoid a double-lookup, I think you could tweak this method to:

```suggestion
  /// Return the instruction with a pointer distance of \p PtrDist, or insert \p  InstIdx
  /// at \p PtrDist if no such instruction exists. Returns the existing instruction, if found,
  /// or std::nullopt on insertion.
  std::optional<unsigned> getOrInsertInstIdx(int PtrDist, int InstIdx) const {
    auto [It, Inserted] = Instrs.insert(PtrDist, InstIdx);
    if (!Inserted)
      return It->second;
    return std::nullopt;
  }
```

Reply:
Yes I had something similar locally, I'll push it today.